source/common/signal/fatal_error_handler.h

```diff
@@ -35,5 +40,31 @@ void removeFatalErrorHandler(const FatalErrorHandlerInterface& handler);
  * called from a fatal signal handler.
  */
 void callFatalErrorHandlers(std::ostream& os);
+
+/**
+ * Creates the singleton FatalActionManager if not already created and
+ * registers the specified actions to run on failure.
+ */
+void registerFatalActions(
+    std::unique_ptr<std::list<const Server::Configuration::FatalActionPtr>> safe_actions,
+    std::unique_ptr<std::list<const Server::Configuration::FatalActionPtr>> unsafe_actions,
+    Envoy::Server::Instance* server);
+
+/**
+ * Tries to run all of the safe fatal actions. Only one thread will
+ * be allowed to run the fatal functions. This returns true if the caller should
+ * continue running the failure functions and have ran the safe actions.
+ * Otherwise it returns false.
+ */
+bool runSafeActions();
+
+/**
+ * Tries to run all of the unsafe fatal actions. Call this only
+ * after calling runSafeActions.
+ *
+ * Returns whether the thread was successfully able to run the
+ * unsafe actions.
+ */
+bool runUnsafeActions();
```

Comment:
Why are these exposed as two separate functions? Do we expect to call them independently?

Reply:
As you suspected, we want to be able to call them independently. 

Envoy's signal handlers can be overwritten by deployment specific handlers, so having these be uncoupled is necessary.